README.markdown
<!--
SPDX-FileCopyrightText: 2018-2022 Joonas Javanainen <joonas.javanainen@gmail.com>

SPDX-License-Identifier: CC0-1.0
-->

# mcp2210-rs: Microchip MCP2210 library

Minimum Rust version: 1.63

[MCP2210 datasheet](http://ww1.microchip.com/downloads/en/devicedoc/22288a.pdf)

[![Build Status](https://github.com/Gekkio/mcp2210-rs/actions/workflows/ci.yml/badge.svg)](https://github.com/Gekkio/mcp2210-rs/actions)
[![Latest release on crates.io](https://img.shields.io/crates/v/mcp2210.svg)](https://crates.io/crates/mcp2210)
[![Documentation on docs.rs](https://docs.rs/mcp2210/badge.svg)](https://docs.rs/mcp2210)

To use `mcp2210`, you'll need to add it and `hidapi` to your dependencies.

```bash
cargo add mcp2210 hidapi
```

### ⚠️ WARNING: This code sends 0xaa55 on the MCP2210's SPI bus. If you have a device connected to the SPI bus, ensure this will not harm it. ⚠️

This code sends 0xaa55 on the MCP2210's SPI bus MOSI pin and asserts that the same data is simultaneously recieved at the MISO pin. The circuit required for this is simply a wire between the MOSI and MISO pins of the MCP2210 and no real slave device.

```rust
use hidapi::HidApi;
use mcp2210::{open_first, Commands, SpiMode, SpiTransferSettings};

fn main() {
    let hidapi_context = HidApi::new().expect("Could not create hidapi context");
    let mut mcp = open_first(&hidapi_context).expect("Failed to connect");
    mcp.set_spi_transfer_settings(&SpiTransferSettings {
        bit_rate: 1_000,
        bytes_per_tx: 2,
        spi_mode: SpiMode::Mode0,
        ..Default::default()
    })
    .expect("Failed to set settings");
    let mut buf = Vec::new();
    mcp.spi_transfer_to_end(&[0xaa, 0x55], &mut buf)
        .expect("SPI transfer failed");
    assert_eq!(buf, [0xaa, 0x55]);
}
```

## License

Licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally
submitted for inclusion in the work by you, as defined in the Apache-2.0
license, shall be dual licensed as above, without any additional terms or
conditions.


CHANGELOG.markdown
<!--
SPDX-FileCopyrightText: 2023 Joonas Javanainen <joonas.javanainen@gmail.com>

SPDX-License-Identifier: CC0-1.0
-->

# Change Log

## [Unreleased]

## 0.2.0 - 2023-12-12

### Added

- `Debug` implementation for `UsbParameters`
- New example: `info`, which dumps all information and state from MCP2210

### Changed

- Improved loopback example
- Improved example code in README
- Switched from libudev + raw hidraw (Linux only) to `hidapi`, which works on macOS and Windows too. **Breaking change**
- Many functions in the API have been redesigned. **Breaking change**
- Bump MSRV to 1.63

## 0.1.0 - 2019-07-17

- Initial release


tests/api_backward_compatibility.rs
// Explicit types are used on variables since consumers of this crate may rely on those types

#[test]
#[ignore = "Only testing that this builds. If a device is connected, we don't know if it's safe to use it."]
fn pre_windows_support_examples() {
    use mcp2210::{Commands, Mcp2210, SpiMode, SpiTransferSettings};

    // The scan_devices_with_filter() function used the libudev::Device type. Consumers of this function
    // will likely need to rewrite the closure.

    {
        #[allow(deprecated)]
        let _ = Mcp2210::open("/dev/hidraw0").expect("Failed to open device");
    }

    {
        // If consumers of this crate depend on either the scan_devices() or scan_devices_with_filter()
        // function returning a Vec<PathBuf>, they will break.
        let devices = mcp2210::scan_devices().expect("Failed to scan devices");
        let device = devices.iter().next().expect("No devices found");
        #[allow(deprecated)]
        let mut mcp: Mcp2210 = Mcp2210::open(&device).expect("Failed to open device");
        mcp.set_spi_transfer_settings(&SpiTransferSettings {
            bit_rate: 1_000,
            bytes_per_tx: 2,
            spi_mode: SpiMode::Mode0,
            ..Default::default()
        })
        .expect("Failed to set settings");

        mcp.get_chip_status().unwrap();
        mcp.get_interrupt_event_counter().unwrap();
        mcp.get_chip_settings().unwrap();
        mcp.get_gpio_direction().unwrap();
        mcp.get_gpio_value().unwrap();
        mcp.get_nvram_spi_transfer_settings().unwrap();
        mcp.get_nvram_chip_settings().unwrap();
        mcp.get_nvram_usb_parameters().unwrap();
        mcp.get_nvram_usb_product_name().unwrap();
        mcp.get_nvram_usb_vendor_name().unwrap();

        let mut buf: Vec<u8> = Vec::new();
        mcp.spi_transfer_to_end(&[0xaa, 0x55], &mut buf)
            .expect("SPI transfer failed");
        assert_eq!(buf.len(), 2);
        println!("0x{:02x} 0x{:02x}", buf[0], buf[1]); // prints 0xaa 0x55
    }
}


src/lib.rs
// SPDX-FileCopyrightText: 2018-2022 Joonas Javanainen <joonas.javanainen@gmail.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

mod cmds;
mod types;
mod utils;

pub use crate::cmds::*;
pub use crate::types::*;

use hidapi::{DeviceInfo, HidApi, HidDevice, HidError, HidResult};
use std::cmp::min;
use std::error::Error;
use std::fmt;

pub const FACTORY_VID: u16 = 0x04d8;
pub const FACTORY_PID: u16 = 0x00de;

#[derive(Debug)]
pub enum Mcp2210Error {
    Hid(HidError),
    CommandCode { expected: u8, actual: u8 },
    SubCommandCode { expected: u8, actual: u8 },
    InvalidResponse(String),
    UnknownErrorCode(u8),
    StringSize(usize),
    PayloadSize(usize),
    TransferStatus(SpiTransferStatus),

    // MCP2210 error codes
    EepromWrite,            // 0xFA
    AccessDenied,           // 0xFB
    AccessRejected,         // 0xFC
    AccessDeniedRetry,      // 0xFD
    Unavailable,            // 0xF7
    Busy,                   // 0xF8
    UnknownCommandCode(u8), // 0xF9
}

impl fmt::Display for Mcp2210Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use crate::Mcp2210Error::*;
        match self {
            Hid(err) => fmt::Display::fmt(err, f),
            CommandCode { expected, actual } => write!(
                f,
                "Invalid command code (expected {:2x}, got {:2x})",
                expected, actual
            ),
            SubCommandCode { expected, actual } => write!(
                f,
                "Invalid sub-command code (expected {:2x}, got {:2x})",
                expected, actual
            ),
            InvalidResponse(response) => write!(f, "Invalid response ({})", response),
            UnknownErrorCode(code) => write!(f, "Unknown error code {:2x}", code),
            StringSize(size) => write!(
                f,
                "String is too long (expected at most 29 UTF-16 encoded u16 values, got {})",
                size
            ),
            PayloadSize(size) => write!(
                f,
                "Payload is too big (expected at most 60 bytes, got {})",
                size
            ),
            TransferStatus(status) => write!(f, "Unexpected SPI transfer status {:?}", status),
            EepromWrite => write!(f, "EEPROM write failure"),
            AccessDenied => write!(f, "Access denied"),
            AccessRejected => write!(f, "Access rejected"),
            AccessDeniedRetry => write!(f, "Access denied, retrying allowed"),
            Unavailable => write!(f, "SPI bus unavailable"),
            Busy => write!(f, "SPI bus busy"),
            UnknownCommandCode(code) => write!(f, "Unknown command code {:2x}", code),
        }
    }
}

impl Error for Mcp2210Error {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        use crate::Mcp2210Error::*;
        match self {
            Hid(err) => Some(err),
            _ => None,
        }
    }
}

const BUFFER_SIZE: usize = 64;

pub type Buffer = [u8; BUFFER_SIZE];

pub const MAX_BIT_RATE: u32 = 12_000_000;

pub struct Mcp2210 {
    device: HidDevice,
}

impl CommandResponse for Mcp2210 {
    fn command_response(&mut self, cmd: &Buffer, res: &mut Buffer) -> HidResult<()> {
        let mut data_to_write = [0; 1 + BUFFER_SIZE];
        data_to_write[0] = 0x00; // HID Report ID. For devices which only support a single report, this must be set to 0x0.
        data_to_write[1..].copy_from_slice(cmd);
        // At this point, length of data_to_write will be 1+BUFFER_SIZE == 65 and responses from the MCP2210 are always
        // BUFFER_SIZE. Therefore, this should only take single reports and these asserts should be good assumptions.
        assert_eq!(self.device.write(&data_to_write)?, data_to_write.len());
        assert_eq!(self.device.read(res)?, BUFFER_SIZE);
        Ok(())
    }
}

impl Mcp2210 {
    /// Converts a HidDevice to a Mcp2210.
    ///
    /// If the passed HidDevice is not actually a MCP2210 device, unexpected things are likely to happen when you
    /// use the Mcp2210 later.
    pub fn new(device: HidDevice) -> Mcp2210 {
        Mcp2210 { device }
    }

    pub fn spi_transfer_to_end(
        &mut self,
        mut data: &[u8],
        buf: &mut Vec<u8>,
    ) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        {
            let len = min(data.len(), 60);
            let res = self.spi_transfer(&data[..len], &mut res)?;
            data = &data[len..];
            if res.status != SpiTransferStatus::Started {
                return Err(Mcp2210Error::TransferStatus(res.status));
            }
        }
        loop {
            let len = min(data.len(), 60);
            match self.spi_transfer(&data[..len], &mut res) {
                Ok(res) => {
                    data = &data[len..];
                    buf.extend(res.data);
                    if res.status == SpiTransferStatus::Finished {
                        break;
                    }
                }
                Err(Mcp2210Error::Busy) => (),
                Err(err) => return Err(err),
            }
        }
        Ok(())
    }
}

/// True if the device has the MCP2210's factory Vendor ID (VID) and Product ID (VID).
pub fn is_mcp2210(device_info: &DeviceInfo) -> bool {
    device_info.vendor_id() == FACTORY_VID && device_info.product_id() == FACTORY_PID
}

/// Open the first HID device it finds with the MCP2210's factory Vendor ID (VID) and Product ID (PID).
///
/// When multiple devices with the MCP2210's factory VID and PID are available, then the first one
/// found in the internal device list will be used. There are however no guarantees, which device this
/// will be.
pub fn open_first(hidapi_context: &HidApi) -> Result<Mcp2210, Mcp2210Error> {
    let mcp = hidapi_context
        .open(FACTORY_VID, FACTORY_PID)
        .map_err(Mcp2210Error::Hid)?;
    Ok(Mcp2210::new(mcp))
}


src/cmds.rs
// SPDX-FileCopyrightText: 2018-2022 Joonas Javanainen <joonas.javanainen@gmail.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use super::{Buffer, Mcp2210Error};
use crate::types::*;
use crate::utils::{as_u16, encode_utf16_to_buffer};
use hidapi::HidResult;
use std::cmp::min;

pub trait CommandResponse {
    fn command_response(&mut self, cmd: &Buffer, res: &mut Buffer) -> HidResult<()>;
    fn do_command<F>(&mut self, cmd_code: u8, res: &mut Buffer, f: F) -> Result<(), Mcp2210Error>
    where
        F: FnOnce(&mut Buffer),
    {
        let mut cmd: Buffer = [0; 64];
        cmd[0] = cmd_code;
        f(&mut cmd);
        self.command_response(&cmd, res)
            .map_err(Mcp2210Error::Hid)?;
        if cmd_code != res[0] {
            return Err(Mcp2210Error::CommandCode {
                expected: cmd_code,
                actual: res[0],
            });
        }
        match res[1] {
            0x00 => Ok(()),
            0xf7 => Err(Mcp2210Error::Unavailable),
            0xf8 => Err(Mcp2210Error::Busy),
            0xf9 => Err(Mcp2210Error::UnknownCommandCode(cmd_code)),
            0xfa => Err(Mcp2210Error::EepromWrite),
            0xfb => Err(Mcp2210Error::AccessDenied),
            0xfc => Err(Mcp2210Error::AccessRejected),
            0xfd => Err(Mcp2210Error::AccessDeniedRetry),
            err_code => Err(Mcp2210Error::UnknownErrorCode(err_code)),
        }
    }
    fn do_sub_command<F>(
        &mut self,
        cmd_code: u8,
        sub_cmd_code: u8,
        res: &mut Buffer,
        f: F,
    ) -> Result<(), Mcp2210Error>
    where
        F: FnOnce(&mut Buffer),
    {
        self.do_command(cmd_code, res, |cmd| {
            cmd[1] = sub_cmd_code;
            f(cmd);
        })?;
        if res[2] != sub_cmd_code {
            Err(Mcp2210Error::SubCommandCode {
                expected: sub_cmd_code,
                actual: res[2],
            })
        } else {
            Ok(())
        }
    }
}

pub trait Commands: CommandResponse {
    fn get_chip_status(&mut self) -> Result<ChipStatus, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x10, &mut res, |_| {})?;
        ChipStatus::from_buffer(&res).map_err(Mcp2210Error::InvalidResponse)
    }
    fn cancel_spi_transfer(&mut self) -> Result<ChipStatus, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x11, &mut res, |_| {})?;
        ChipStatus::from_buffer(&res).map_err(Mcp2210Error::InvalidResponse)
    }
    fn get_interrupt_event_counter(&mut self) -> Result<u16, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x12, &mut res, |cmd| {
            cmd[1] = 0xff;
        })?;
        Ok(as_u16(res[4], res[5]))
    }
    fn reset_interrupt_event_counter(&mut self) -> Result<u16, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x12, &mut res, |cmd| {
            cmd[1] = 0x00;
        })?;
        Ok(as_u16(res[4], res[5]))
    }
    fn get_chip_settings(&mut self) -> Result<ChipSettings, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x20, &mut res, |_| {})?;
        ChipSettings::from_buffer(&res).map_err(Mcp2210Error::InvalidResponse)
    }
    fn set_chip_settings(&mut self, settings: &ChipSettings) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x21, &mut res, |cmd| {
            settings.write_to_buffer(cmd);
        })
    }
    fn set_gpio_value(&mut self, value: GpioValue) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x30, &mut res, |cmd| {
            let value = value.bits();
            cmd[4] = value as u8;
            cmd[5] = (value >> 8) as u8;
        })
    }
    fn get_gpio_value(&mut self) -> Result<GpioValue, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x31, &mut res, |_| {})?;
        Ok(GpioValue::from_bits_truncate(as_u16(res[4], res[5])))
    }
    fn set_gpio_direction(&mut self, direction: GpioDirection) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x32, &mut res, |cmd| {
            let direction = direction.bits();
            cmd[4] = direction as u8;
            cmd[5] = (direction >> 8) as u8;
        })
    }
    fn get_gpio_direction(&mut self) -> Result<GpioDirection, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x33, &mut res, |_| {})?;
        Ok(GpioDirection::from_bits_truncate(as_u16(res[4], res[5])))
    }
    fn set_spi_transfer_settings(
        &mut self,
        settings: &SpiTransferSettings,
    ) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x40, &mut res, |cmd| {
            settings.write_to_buffer(cmd);
        })
    }
    fn get_spi_transfer_settings(&mut self) -> Result<SpiTransferSettings, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x41, &mut res, |_| {})?;
        SpiTransferSettings::from_buffer(&res).map_err(Mcp2210Error::InvalidResponse)
    }
    fn spi_transfer<'a>(
        &mut self,
        data: &[u8],
        res: &'a mut Buffer,
    ) -> Result<SpiTransferResponse<'a>, Mcp2210Error> {
        if data.len() > 60 {
            return Err(Mcp2210Error::PayloadSize(data.len()));
        }
        let mosi_len = min(data.len(), 60);
        self.do_command(0x42, res, |cmd| {
            cmd[1] = mosi_len as u8;
            cmd[4..][..mosi_len].copy_from_slice(&data[..mosi_len]);
        })?;
        let miso_len = res[2] as usize;
        Ok(SpiTransferResponse {
            data: &res[4..][..miso_len],
            status: SpiTransferStatus::from_u8(res[3]).map_err(|v| {
                Mcp2210Error::InvalidResponse(format!("Invalid SPI transfer status: {:02x}", v))
            })?,
        })
    }
    fn read_eeprom(&mut self, addr: u8) -> Result<u8, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x50, &mut res, |cmd| {
            cmd[1] = addr;
        })?;
        if res[2] != addr {
            return Err(Mcp2210Error::InvalidResponse(format!(
                "Invalid EEPROM address {:2x}",
                addr
            )));
        }
        Ok(res[3])
    }
    fn write_eeprom(&mut self, addr: u8, data: u8) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x51, &mut res, |cmd| {
            cmd[1] = addr;
            cmd[2] = data;
        })
    }
    fn set_nvram_spi_transfer_settings(
        &mut self,
        settings: &SpiTransferSettings,
    ) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_sub_command(0x60, 0x10, &mut res, |cmd| {
            settings.write_to_buffer(cmd);
        })
    }
    fn set_nvram_chip_settings(
        &mut self,
        settings: &ChipSettings,
        password: Option<&[u8; 8]>,
    ) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_sub_command(0x60, 0x20, &mut res, |cmd| {
            settings.write_to_buffer(cmd);
            if let Some(password) = password {
                cmd[19..27].copy_from_slice(password);
            }
        })
    }
    fn set_nvram_usb_parameters(&mut self, params: &UsbParameters) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_sub_command(0x60, 0x30, &mut res, |cmd| {
            params.write_to_buffer(cmd);
        })
    }
    fn set_nvram_usb_product_name(&mut self, name: &str) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        let size = name.encode_utf16().count();
        if size > 29 {
            return Err(Mcp2210Error::StringSize(size));
        }
        self.do_sub_command(0x60, 0x40, &mut res, |cmd| {
            cmd[4] = (size as u8) * 2 + 2;
            cmd[5] = 0x03;
            encode_utf16_to_buffer(name, &mut cmd[6..]);
        })
    }
    fn set_nvram_usb_vendor_name(&mut self, name: &str) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        let size = name.encode_utf16().count();
        if size > 29 {
            return Err(Mcp2210Error::StringSize(size));
        }
        self.do_sub_command(0x60, 0x50, &mut res, |cmd| {
            cmd[4] = (size as u8) * 2 + 2;
            cmd[5] = 0x03;
            encode_utf16_to_buffer(name, &mut cmd[6..]);
        })
    }
    fn get_nvram_spi_transfer_settings(&mut self) -> Result<SpiTransferSettings, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_sub_command(0x61, 0x10, &mut res, |_| {})?;
        SpiTransferSettings::from_buffer(&res).map_err(Mcp2210Error::InvalidResponse)
    }
    fn get_nvram_chip_settings(&mut self) -> Result<ChipSettings, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_sub_command(0x61, 0x20, &mut res, |_| {})?;
        ChipSettings::from_buffer(&res).map_err(Mcp2210Error::InvalidResponse)
    }
    fn get_nvram_usb_parameters(&mut self) -> Result<UsbParameters, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_sub_command(0x61, 0x30, &mut res, |_| {})?;
        UsbParameters::from_buffer(&res).map_err(Mcp2210Error::InvalidResponse)
    }
    fn get_nvram_usb_product_name(&mut self) -> Result<String, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_sub_command(0x61, 0x40, &mut res, |_| {})?;
        let str_bytes = (res[4] - 2) as usize;
        let str_chars = str_bytes / 2;
        let mut char_buf = [0; 29];
        for (idx, chunk) in res[6..][..str_bytes].chunks(2).enumerate() {
            char_buf[idx] = as_u16(chunk[0], chunk[1]);
        }
        Ok(String::from_utf16_lossy(&char_buf[..str_chars]))
    }
    fn get_nvram_usb_vendor_name(&mut self) -> Result<String, Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_sub_command(0x61, 0x50, &mut res, |_| {})?;
        let str_bytes = (res[4] - 2) as usize;
        let str_chars = str_bytes / 2;
        let mut char_buf = [0; 29];
        for (idx, chunk) in res[6..][..str_bytes].chunks(2).enumerate() {
            char_buf[idx] = as_u16(chunk[0], chunk[1]);
        }
        Ok(String::from_utf16_lossy(&char_buf[..str_chars]))
    }
    fn send_access_password(&mut self, password: &[u8; 8]) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x70, &mut res, |cmd| {
            cmd[4..11].copy_from_slice(password);
        })
    }
    fn request_bus_release(&mut self, ack_value: bool) -> Result<(), Mcp2210Error> {
        let mut res: Buffer = [0; 64];
        self.do_command(0x80, &mut res, |cmd| {
            cmd[1] = if ack_value { 0x01 } else { 0x00 };
        })
    }
}

impl<T> Commands for T where T: CommandResponse {}

#[cfg(test)]
struct TestTx {
    cmd: Buffer,
    res: Buffer,
}

#[cfg(test)]
impl TestTx {
    fn new(res: &[u8]) -> TestTx {
        assert!(res.len() <= 64);
        let mut tx = TestTx {
            cmd: [0; 64],
            res: [0; 64],
        };
        tx.res[..res.len()].copy_from_slice(res);
        tx
    }
}

#[cfg(test)]
impl CommandResponse for TestTx {
    fn command_response(&mut self, cmd: &Buffer, res: &mut Buffer) -> HidResult<()> {
        self.cmd.copy_from_slice(cmd);
        res.copy_from_slice(&self.res);
        Ok(())
    }
}

#[test]
fn test_get_chip_status() {
    let mut tx = TestTx::new(&[0x10, 0x00, 0x01, 0x02, 42, 0x01]);
    let status = tx.get_chip_status().unwrap();
    let mut expected_cmd = [0; 64];
    expected_cmd[0] = 0x10;
    assert_eq!(tx.cmd.as_ref(), expected_cmd.as_ref());
    assert!(!status.is_bus_release_pending);
    assert_eq!(status.bus_owner, BusOwner::ExternalMaster);
    assert_eq!(status.password_attempt_count, 42);
    assert!(status.is_password_guessed);
}

#[test]
fn test_cancel_spi_transfer() {
    let mut tx = TestTx::new(&[0x11, 0x00, 0x00, 0x01, 79, 0x00]);
    let status = tx.cancel_spi_transfer().unwrap();
    let mut expected_cmd = [0; 64];
    expected_cmd[0] = 0x11;
    assert_eq!(tx.cmd.as_ref(), expected_cmd.as_ref());
    assert!(status.is_bus_release_pending);
    assert_eq!(status.bus_owner, BusOwner::UsbBridge);
    assert_eq!(status.password_attempt_count, 79);
    assert!(!status.is_password_guessed);
}


src/utils.rs
// SPDX-FileCopyrightText: 2018-2022 Joonas Javanainen <joonas.javanainen@gmail.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

#[inline]
pub fn as_bool(v: u8) -> Result<bool, u8> {
    match v {
        0x00 => Ok(false),
        0x01 => Ok(true),
        _ => Err(v),
    }
}

#[inline]
pub fn as_u16(a: u8, b: u8) -> u16 {
    (u16::from(b) << 8) | u16::from(a)
}

#[inline]
pub fn as_u32(a: u8, b: u8, c: u8, d: u8) -> u32 {
    (u32::from(d) << 24) | (u32::from(c) << 16) | (u32::from(b) << 8) | u32::from(a)
}

#[inline]
pub fn encode_utf16_to_buffer(s: &str, buf: &mut [u8]) {
    let mut idx = 0;
    for ch in s.encode_utf16() {
        buf[idx] = ch as u8;
        buf[idx + 1] = (ch >> 8) as u8;
        idx += 2;
    }
}

#[test]
fn test_as_u16() {
    assert_eq!(as_u16(0xaa, 0x55), 0x55aa);
}

#[test]
fn test_as_u32() {
    assert_eq!(as_u32(0x11, 0x22, 0x33, 0x44), 0x44332211);
}

#[test]
fn test_encode_utf16_to_buffer() {
    let mut buf = [0; 14];
    encode_utf16_to_buffer("보라색 고양이", &mut buf);
    assert_eq!(
        buf,
        [0xf4, 0xbc, 0x7c, 0xb7, 0xc9, 0xc0, 0x20, 0x00, 0xe0, 0xac, 0x91, 0xc5, 0x74, 0xc7]
    );
}


examples/info.rs
// SPDX-FileCopyrightText: 2019-2022 Jonas Schievink <jonasschievink@gmail.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

//! This example reads out all information and state from the MCP2210 and writes it to the console.

extern crate mcp2210;

use hidapi::HidApi;
use mcp2210::{open_first, Commands};

fn main() {
    let hidapi_context = HidApi::new().expect("Could not create hidapi context");
    let mut mcp = open_first(&hidapi_context).expect("Failed to connect");
    println!("Current Chip Status");
    println!("===================");
    println!(
        "{:#?}",
        mcp.get_chip_status().expect("Failed to read chip status")
    );
    println!(
        "Interrupt event counter: {}",
        mcp.get_interrupt_event_counter()
            .expect("Failed to read interrupt count")
    );
    println!(
        "{:#?}",
        mcp.get_chip_settings()
            .expect("Failed to read chip settings")
    );
    println!(
        "GPIO directions (inputs): {:#?}",
        mcp.get_gpio_direction()
            .expect("Failed to read GPIO directions")
    );
    println!(
        "GPIO values: {:#?}",
        mcp.get_gpio_value().expect("Failed to read GPIO values")
    );
    println!();
    println!("NVRAM settings");
    println!("==============");
    println!(
        "{:#?}",
        mcp.get_nvram_spi_transfer_settings()
            .expect("Failed to read NVRAM SPI transfer settings")
    );
    println!(
        "{:#?}",
        mcp.get_nvram_chip_settings()
            .expect("Failed to read NVRAM chip settings")
    );
    println!(
        "{:#?}",
        mcp.get_nvram_usb_parameters()
            .expect("Failed to read NVRAM USB parameters")
    );
    println!(
        "Product name: {:?}",
        mcp.get_nvram_usb_product_name()
            .expect("Failed to read NVRAM USB product name")
    );
    println!(
        "Vendor name: {:?}",
        mcp.get_nvram_usb_vendor_name()
            .expect("Failed to read NVRAM USB vendor name")
    );
}


examples/loopback.rs
// SPDX-FileCopyrightText: 2018-2022 Joonas Javanainen <joonas.javanainen@gmail.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

extern crate mcp2210;

use hidapi::HidApi;
use mcp2210::{open_first, Commands, SpiMode, SpiTransferSettings};

fn main() {
    //! ##################################################################################
    //! ##                                ⚠️ WARNING ⚠️                                   ##
    //! ## This code sends 0xaa55 on the MCP2210's SPI bus.                             ##
    //! ## If you have a device connected to the SPI bus, ensure this will not harm it. ##
    //! ##################################################################################
    //! #
    //! This code sends 0xaa55 on the MCP2210's SPI bus MOSI pin and asserts that the same
    //! data is simultaneously recieved at the MISO pin. The circuit required for this is
    //! simply a wire between the MOSI and MISO pins of the MCP2210 and no real slave device.

    let hidapi_context = HidApi::new().expect("Could not create hidapi context");
    let mut mcp = open_first(&hidapi_context).expect("Failed to connect");
    mcp.set_spi_transfer_settings(&SpiTransferSettings {
        bit_rate: 1_000,
        bytes_per_tx: 2,
        spi_mode: SpiMode::Mode0,
        ..Default::default()
    })
    .expect("Failed to set settings");
    let mut buf = Vec::new();
    mcp.spi_transfer_to_end(&[0xaa, 0x55], &mut buf)
        .expect("SPI transfer failed");
    assert_eq!(buf, [0xaa, 0x55]);
}
